webapp/src/components/App/Studies/StudyTree/index.tsx

```diff
@@ -0,0 +1,298 @@
+  if (pathParts.length < 2) {
+    return [];
+  }
+  // path parts should be ["root", workspace, "folder1", ...]
+  const workspace = pathParts[1];
+  const subPath = pathParts.slice(2).join("/");
+  return await api.getFolders(workspace, subPath);
+}
+
+/**
+ * Fetch and insert the subfolders under the given paths into the study tree.
+ *
+ * This function is used to fill the study tree when the user clicks on a folder.
+ *
+ * Subfolders are inserted only if they don't exist already in the tree.
+ *
+ * @param paths list of paths to fetch the subfolders for
+ * @param studiesTree study tree to insert the subfolders into
+ * @returns study tree with the subfolders inserted if they weren't already there.
+ */
+async function fetchAndInsertSubfolders(
+  paths: string[],
+  studiesTree: StudyTreeNode,
+): Promise<[StudyTreeNode, string[]]> {
+  return paths.reduce(
+    async (acc, path) => {
+      const [accTree, accFailedPaths] = await acc;
+      try {
+        const subfolders = await fetchSubfolders(path);
+        return [insertFoldersIfNotExist(accTree, subfolders), accFailedPaths];
+      } catch (error) {
+        console.error("failed to load path ", path, error);
+        return [accTree, [...accFailedPaths, path]];
+      }
+    },
+    Promise.resolve<[StudyTreeNode, string[]]>([studiesTree, []]),
+  );
+}
+
+/**
+ * Insert a workspace into the study tree if it doesn't exist already.
+ *
+ * @param workspace key of the workspace
+ * @param stydyTree study tree to insert the workspace into
+ * @returns study tree with the empty workspace inserted if it wasn't already there.
+ */
+function insertWorkspaceIfNotExist(
+  stydyTree: StudyTreeNode,
+  workspace: string,
+) {
+  const emptyNode = { name: workspace, path: `/${workspace}`, children: [] };
+  if (stydyTree.children.some((child) => child.name === workspace)) {
+    return stydyTree;
+  }
+  return {
+    ...stydyTree,
+    children: [...stydyTree.children, emptyNode],
+  };
+}
+
+/**
+ * Insert several workspaces into the study tree if they don't exist already in the tree.
+ *
+ * @param workspaces workspaces to insert into the tree
+ * @param stydyTree study tree to insert the workspaces into
+ * @returns study tree with the empty workspaces inserted if they weren't already there.
+ */
+export function insertWorkspacesIfNotExist(
+  stydyTree: StudyTreeNode,
+  workspaces: string[],
+): StudyTreeNode {
+  return workspaces.reduce((acc, workspace) => {
+    return insertWorkspaceIfNotExist(acc, workspace);
+  }, stydyTree);
+}
+
+/**
+ * fetch and insert the workspaces into the study tree.
+ * @param studyTree study tree to insert the workspaces into
+ * @returns study tree with the workspaces inserted if they weren't already there.
+ */
+async function fetchAndInsertWorkspaces(
+  studyTree: StudyTreeNode,
+): Promise<StudyTreeNode> {
+  const workspaces = await api.getWorkspaces();
+  return insertWorkspacesIfNotExist(studyTree, workspaces);
+}
```

Comment:
you can remove intermediate functions, and streamline workspace insertion
```typescript
async function fetchAndInsertWorkspaces(
  studyTree: StudyTreeNode,
): Promise<StudyTreeNode> {
  const workspaces = await api.getWorkspaces();

  return workspaces.reduce((tree, workspace) => {
    const exists = tree.children.some((child) => child.name === workspace);
    if (exists) {
      return tree;
    }

    return {
      ...tree,
      children: [
        ...tree.children,
        { name: workspace, path: `/${workspace}`, children: [] },
      ],
    };
  }, studyTree);
}
```

Reply:
I want to write the unit test for insertWorkspacesIfNotExist it's easier if it's not coupled with the fetch

Reply:
ok 👍🏼